Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -21,981 +21,660 @@
-        return 0;
-    }
-
-    rect = CGRectIntegral(rect);
+CGImageRef CGImageCreate(size_t width, size_t height, size_t bitsPerComponent,
+                         size_t bitsPerPixel, size_t bytesPerRow,
+                         CGColorSpaceRef colorSpace, CGBitmapInfo bitmapInfo,
+                         CGDataProviderRef provider, const float *decode,
+                         bool shouldInterpolate,
+                         CGColorRenderingIntent intent) {
 
-    CGRect imgRefSize;
-    imgRefSize.origin.x = 0;
-    imgRefSize.origin.y = 0;
-    imgRefSize.size.width = (float)ref->Backing()->Width();
-    imgRefSize.size.height = (float)ref->Backing()->Height();
+  RETURN_NULL_IF(((provider == nullptr) ||
+                  ![(NSObject *)provider isKindOfClass:[NSData class]]));
 
-    rect = CGRectIntersection(rect, imgRefSize);
+  NSData *dataProvider = (__bridge NSData *)provider;
 
-    __CGSurfaceInfo surfaceInfo;
-    ref->Backing()->GetSurfaceInfoWithoutPixelPtr(&surfaceInfo);
+  unsigned char *data = (unsigned char *)[dataProvider bytes];
 
-    // Override width and height with the rect
-    surfaceInfo.width = rect.size.width;
-    surfaceInfo.height = rect.size.height;
+  bool colorSpaceAllocated = false;
+  if (colorSpace == NULL) {
+    if (bytesPerRow >= (width * 3)) {
+      TraceWarning(
+          TAG,
+          L"Warning: colorSpace = NULL, assuming RGB based on bytesPerRow.");
+      colorSpace = CGColorSpaceCreateDeviceRGB();
+    } else {
+      TraceWarning(
+          TAG,
+          L"Warning: colorSpace = NULL, assuming Gray based on bytesPerRow.");
+      colorSpace = CGColorSpaceCreateDeviceGray();
+    }
 
-    assert(surfaceInfo.surfaceData == NULL);
+    colorSpaceAllocated = true;
+  }
 
-    CGImageRef newImage = new CGBitmapImage(surfaceInfo);
+  ComPtr<IWICBitmap> image;
+  ComPtr<IWICImagingFactory> imageFactory = _GetWICFactory();
 
-    int startX = (int)rect.origin.x;
-    int startY = (int)rect.origin.y;
-    int sizeX = (int)rect.size.width;
-    int sizeY = (int)rect.size.height;
+  REFGUID pixelFormat = _CGImageGetWICPixelFormat(
+      bitsPerComponent, bitsPerPixel, colorSpace, bitmapInfo);
 
-    BYTE* srcIn =
-        ((BYTE*)ref->Backing()->LockImageData()) + startY * ref->Backing()->BytesPerRow() + startX * ref->Backing()->BytesPerPixel();
-    BYTE* destOut = (BYTE*)newImage->Backing()->LockImageData();
+  // TODO: take care of the stride (mapping of the guids)
+  HRESULT status = imageFactory->CreateBitmapFromMemory(
+      width, height, pixelFormat, bytesPerRow, height * bytesPerRow, data,
+      &image);
 
-    for (int curY = 0; curY < sizeY; curY++) {
-        memmove(destOut, srcIn, newImage->Backing()->BytesPerRow());
+  CGImageRef imageRef = __CGImage::CreateInstance();
+  imageRef->ImageSource().Attach(image.Detach());
 
-        srcIn += ref->Backing()->BytesPerRow();
-        destOut += newImage->Backing()->BytesPerRow();
-    }
+  imageRef->SetBitmapInfo(bitmapInfo)
```

Comment:
It seems much clearer given these. 
We are already have inline functions for getters.

---
In reply to: [84331527](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84331527) [](ancestors = 84331527)